.azure-pipelines-templates/matrix.yml

```diff
@@ -1,116 +1,170 @@
+    Build_Release:
+      container: sgx
+      pool: Ubuntu-1804-DC4s
+      dependsOn: ['Test_NoSGX_CFT_Debug', 'Test_SGX_CFT_Debug', 'Test_NoSGX_CFT_SAN', 'Test_NoSGX_BFT_Debug', 'Test_SGX_BFT_Debug', 'Test_NoSGX_BFT_SAN']
     Release:
       container: sgx
       pool: Ubuntu-1804-DC4s
-      dependsOn: ['BuildAndTest_NoSGX_CFT_Debug', 'BuildAndTest_SGX_CFT_Debug', 'BuildAndTest_NoSGX_CFT_SAN', 'BuildAndTest_NoSGX_BFT_Debug', 'BuildAndTest_SGX_BFT_Debug', 'BuildAndTest_NoSGX_BFT_SAN']
+      dependsOn: ['Build_SGX_Release']
     Metrics:
       container: nosgx
       pool: Ubuntu-1804-D8s_v3
-      dependsOn: ['BuildAndTest_SGX_CFT_Perf', 'BuildAndTest_SGX_BFT_Perf']
+      dependsOn: ['Test_SGX_CFT_Perf', 'Test_SGX_BFT_Perf']
     Publish:
       container: nosgx
       pool: Ubuntu-1804-D8s_v3
-      dependsOn: ['BuildAndTest_SGX_CFT_Release', 'BuildAndTest_SGX_BFT_Release']
-    Perf:
+      dependsOn: ['Test_SGX_CFT_Release', 'Test_SGX_BFT_Release']
+    Build_Perf:
       pool: CCF-Perf
       dependsOn:  ['Formatting_and_License_Checks']
+    Perf:
+      pool: CCF-Perf
+      dependsOn:  ['Build_SGX_Perf']
 
   build:
-    common:
-      cmake_args: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache"'
-    NoSGX:
-      cmake_args: '-DTARGET=virtual -DCOVERAGE=ON'
-    SGX:
-      cmake_args: ''
-    BFT:
-      cmake_args: '-DPBFT=ON'
-    CFT:
-      cmake_args: ''
-    debug:
-      cmake_args: '-DCMAKE_BUILD_TYPE=Debug -DVERBOSE_LOGGING=ON -DBUILD_SMALLBANK=OFF'
-    install:
-      install_prefix: '/tmp/ccf-install'
-      cmake_args: '-DCMAKE_INSTALL_PREFIX=/tmp/ccf-install'
-    san:
-      cmake_args: '-DSAN=ON'
-    perf:
-      cmake_args: '-DVERBOSE_LOGGING=ON -DDISTRIBUTE_PERF_TESTS="`../.nodes.sh`"'
+    Build_SGX: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache" -DCMAKE_BUILD_TYPE=Debug -DVERBOSE_LOGGING=ON -DBUILD_SMALLBANK=OFF'
+    Build_NoSGX: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache" -DCMAKE_BUILD_TYPE=Debug -DVERBOSE_LOGGING=ON -DBUILD_SMALLBANK=OFF -DTARGET=virtual -DCOVERAGE=ON'
+    SGX: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache" -DCMAKE_BUILD_TYPE=Debug -DVERBOSE_LOGGING=ON -DBUILD_SMALLBANK=OFF'
+    NoSGX: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache" -DCMAKE_BUILD_TYPE=Debug -DVERBOSE_LOGGING=ON -DBUILD_SMALLBANK=OFF -DTARGET=virtual -DCOVERAGE=ON'
+    san: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache" -DSAN=ON -DTARGET=virtual -DCOVERAGE=ON'
+    perf: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache" -DVERBOSE_LOGGING=ON -DDISTRIBUTE_PERF_TESTS="`../.nodes.sh`"'
+    install: '-DCMAKE_C_COMPILER_LAUNCHER="ccache" -DCMAKE_CXX_COMPILER_LAUNCHER="ccache" -DCMAKE_INSTALL_PREFIX=/tmp/ccf-install -DCMAKE_INSTALL_PREFIX=/tmp/ccf-install'
+
```

Comment:
I know that these here have a lot of duplication but alongside with the new `Build_SGX` and `Build_NoSGX` params that are needed for the `build_target`s there ended up being duplication in the `cmake_args` we need to pass once for the build and once for the test configuration and this just seems neater.

Reply:
I don't understand why this is denormalised and why we need the Build_SGX and Build_NoSGX build_targets, which seem to duplicate NoSGX and SGX. A lot of this fit nicely into common.